src/circularbuffer.jl

```diff
@@ -0,0 +1,61 @@
+
+"""
+New items are pushed to the back of the list, overwriting values in a circular fashion.
+"""
+type CircularBuffer{T} <: AbstractVector{T}
+    capacity::Int
+    first::Int
+    buffer::Vector{T}
+
+    CircularBuffer(capacity::Int) = new(capacity, 1, T[])
+end
+
+function _buffer_index(cb::CircularBuffer, i::Int)
+    n = length(cb)
+    if i < 1 || i > n
+        error("CircularBuffer out of range. cb=$cb i=$i")
```

Comment:
The lowest common denominator is to `throw(BoundsError())`, which should work on 0.3 or 0.4. But maybe we can just drop 0.3 support here now that 0.4 is released? I just opened #145 to see how others feel...